### PR TITLE
Fix cosmetic filters first party machinery #10035

### PR DESCRIPTION
**Pull request by atuchin-m**
Resolves brave/brave-browser#17906

`ParseDomain()` in `cosmetic_filters.ts` is replaced by a new native function `cf_worker.isFirstPartyUrl()` which calls `net::registry_controlled_domains::SameDomainOrHost()`. This solves the performance issue with building prefix tree.
The problem with returning false for the first party node is resolved.
CosmeticFilteringProtect1p enabled and improved.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by iefremov** [comment]
I don't think a formal security review is need, but @fmarier can you please take a quick look?

File: browser/brave_shields/ad_block_service_browsertest.cc
Comment (iefremov) [inline browser/brave_shields/ad_block_service_browsertest.cc:1725]: I would explicitly explain in a comment why appspot.com is meaningful here

**Review by iefremov — APPROVED** [review]

File: browser/brave_shields/ad_block_service_browsertest.cc
Comment (atuchin-m) [reply browser/brave_shields/ad_block_service_browsertest.cc:1727]: Done.

File: browser/brave_shields/ad_block_service_browsertest.cc
Comment (iefremov) [inline browser/brave_shields/ad_block_service_browsertest.cc:1726]: It's

Comment (atuchin-m) [reply browser/brave_shields/ad_block_service_browsertest.cc]: Fixed

**Comment by iefremov** [comment]
@pes10k can you please sign off these changes? AFAIU with this change we will allow way more elements that before, because `isFirstParty` was essentially broken and used to always return false.
So perhaps the impact would be pretty much user visible?

**Comment by pes10k** [comment]
I am fine with the changes, will approve.  But as long we have a quick follow up issue to look into the possible issues @atuchin-m was unsure about around the checks for text-ads

**Comment by pes10k** [comment]
Before hitting merge though, can someone with a build of this branch confirm that these tests pass? https://dev-pages.brave.software/filtering/text-ads.html

**Review by pes10k — APPROVED** [review]
LGTM as long as https://dev-pages.brave.software/filtering/text-ads.html pass

**Comment by iefremov** [comment]
Thanks @pes10k !

File: components/cosmetic_filters/resources/data/content_cosmetic.ts
Comment (antonok-edm) [inline components/cosmetic_filters/resources/data/content_cosmetic.ts:12]: I believe we can remove our `parse-domain` dependency from `package.json`, this is the only place it's used

**Review by fmarier — APPROVED** [review]

File: components/cosmetic_filters/resources/data/content_cosmetic.ts
Comment (atuchin-m) [reply components/cosmetic_filters/resources/data/content_cosmetic.ts:12]: Removed, thanks!

**Review by antonok-edm — APPROVED** [review]

**Comment by atuchin-m** [comment]
> LGTM as long as https://dev-pages.brave.software/filtering/text-ads.html pass

I checked the tests, nothing has been changed.

**Pull request merged**
